### PR TITLE
Add rule S6441: Unused methods of React components should be removed

### LINKED ISSUE #3214: Add rule S6441: Unused methods of React components should be removed (`react/no-unused-class-component-methods`)
[RSPEC PR](https://github.com/SonarSource/rspec/pull/1096)

### PR DESCRIPTION
**Pull request by victor-diez-sonarsource**
Fixes #3214

### CODE REVIEW

File: eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import { rule as detectReact } from '../utils/rule-detect-react';
+import { rules as reactRules } from 'eslint-plugin-react';
+import { interceptReport, mergeRules } from '../utils';
+
+const noUnusedClassComponentMethod = reactRules['no-unused-class-component-methods'];
+
+export const rule: Rule.RuleModule = {
+  meta: {
+    messages: {
+      unused: 'Method or property "{{name}}" is not used inside component body',
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts:32]: ha, in fact we can do whatever we want with the message, should we try to rephrase it then?
e.g. `Remove this property or method or refactor this component "{{className}}", as "{{name}}" is not used inside component body`

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html [hunk 2]
```diff
@@ -0,0 +1,31 @@
+<p>Methods that are never executed are dead code: unnecessary, inoperative code that should be removed. Cleaning out dead code decreases the size of
+the maintained codebase, making it easier to understand the program and preventing bugs from being introduced.</p>
+<p>This rule reports the methods of React component classes which are neither used nor React lifecycle methods. Note that this rule checks only the
+scope of a component class, which means that issues will be raised on the methods that are only used outside of the class.</p>
```
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html:4]: We should add some phrase here that it's considered a bad practice to use component methods outside of it.

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html (hunk 2 above)
Comment (victor-diez-sonarsource) [reply javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html:4]: but we add that to rspec repo, right?

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    return &lt;h1&gt;{ props.name || getDefaultName() }&lt;/h1&gt;;
+  }
+
+  getDefaultName() {
+    return 'John Smith';
+  }
+}
+</pre>
+<h2>See</h2>
+<p>FIXME: A list of links</p>
```
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html:30]: oops :)

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html (hunk 2 above)
Comment (vilchik-elena) [reply javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6441.html:4]: regenerate the file to have your changes in the PR

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

**Review by vilchik-elena — APPROVED** [review]
LGTM. I would just change text for the link

**Review by saberduck — DISMISSED** [review]

File: eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  meta: {
+    messages: {
+      unused:
+        'Remove this property or method or refactor this component, as "{{name}}" is not used inside component body',
+      unusedWithClass:
+        'Remove this property or method or refactor "{{className}}", as "{{name}}" is not used inside component body',
+    },
+  },
+  create(context: Rule.RuleContext) {
+    function overrideContext(overrides: any) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts:39]: I would move it to the top of the file after the rule definition to keep the logic here cleaner

File: eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    let isReact = false;
+
+    const detectReactContext = overrideContext({
+      report(_descriptor: Rule.ReportDescriptor): void {
+        isReact = true;
+      },
+    });
+
+    // We may need to add deprecated API that the react plugin still relies on if the rule is decorated by
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts:52]: I don't understand this comment

File: eslint-bridge/tests/rules/decorators/sonar-no-unused-class-component-methods.test.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import { RuleTester } from 'eslint';
+import { rule } from 'rules/sonar-no-unused-class-component-methods';
+
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2018, sourceType: 'module' } });
+
+ruleTester.run(`Function parameters with default values should be last`, rule, {
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/decorators/sonar-no-unused-class-component-methods.test.ts:25]: this is copy paste from another rule

File: eslint-bridge/tests/rules/decorators/sonar-no-unused-class-component-methods.test.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import { RuleTester } from 'eslint';
+import { rule } from 'rules/sonar-no-unused-class-component-methods';
+
+const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 2018, sourceType: 'module' } });
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/decorators/sonar-no-unused-class-component-methods.test.ts:23]: as we don't use decorator now, can we rely on comment-based tests?

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts (hunk 5 above)
Comment (victor-diez-sonarsource) [reply eslint-bridge/src/rules/sonar-no-unused-class-component-methods.ts:52]: copy paste from Francois, I will remove

**Pull request merged**